run-build-functions.sh

```diff
@@ -132,32 +139,38 @@ restore_node_modules() {
 
 run_yarn() {
   yarn_version=$1
-  if [ -d $NETLIFY_CACHE_DIR/yarn ]
-  then
-    export PATH=$NETLIFY_CACHE_DIR/yarn/bin:$PATH
-  fi
   restore_home_cache ".yarn_cache" "yarn cache"
 
-  if [ $(which yarn) ] && [ "$(yarn --version)" != "$yarn_version" ]
-  then
-    echo "Found yarn version ($(yarn --version)) that doesn't match expected ($yarn_version)"
-    rm -rf $NETLIFY_CACHE_DIR/yarn $HOME/.yarn
-    npm uninstall yarn -g
-  fi
+  if ! [ $(which corepack) ]; then
+    if [ -d $NETLIFY_CACHE_DIR/yarn ]
+    then
+      export PATH=$NETLIFY_CACHE_DIR/yarn/bin:$PATH
+    fi
 
-  if ! [ $(which yarn) ]
-  then
-    echo "Installing yarn at version $yarn_version"
-    rm -rf $HOME/.yarn
-    bash /usr/local/bin/yarn-installer.sh --version $yarn_version
-    mv $HOME/.yarn $NETLIFY_CACHE_DIR/yarn
-    export PATH=$NETLIFY_CACHE_DIR/yarn/bin:$PATH
+    if [ $(which yarn) ] && [ "$(yarn --version)" != "$yarn_version" ]; then
+        echo "Found yarn version ($(yarn --version)) that doesn't match expected ($yarn_version)"
+        rm -rf $NETLIFY_CACHE_DIR/yarn $HOME/.yarn
+        npm uninstall yarn -g
+    fi
+
+    if ! [ $(which yarn) ]; then
+      echo "Installing yarn at version $yarn_version"
+      rm -rf $HOME/.yarn
+      bash /usr/local/bin/yarn-installer.sh --version $yarn_version
+      mv $HOME/.yarn $NETLIFY_CACHE_DIR/yarn
+      export PATH=$NETLIFY_CACHE_DIR/yarn/bin:$PATH
+    fi
+  else
+    # if corepack is installed use it for changing the yarn version
+    if [ "$(yarn --version)" != "$yarn_version" ]; then
+      echo "Installing yarn at version $yarn_version"
+      corepack prepare yarn@$yarn_version --activate
+    fi
```

Comment:
Could be farteched but I'm thinking on the following scenario:
- Someone running node configured via `NODE_VERSION`, `v18.x` for example (they have corepack).
- This change lands in prod
- `which corepack` will return a non-error exit code
- `yarn` binary won't exist so `yarn --version` will blow up? 🤔
Would this be possible? Should we check for the presence of the yarn executable before too? If so maybe we could move the `which yarn` assertion to the top and use that across the function via a `yarn_exists` flag maybe?

Reply:
That's not the case corepack installs yarn for each node version which is super awesome:

you can test it by: 

1. current node version - uninstall your current yarn binary
2. run `which yarn` (should fail)
3. run `corepack enable`
4. run `corepack prepare yarn@1.22.19 --activate
5. run `which yarn` should work
6. run `nvm install 18` (has corepack but no yarn installed)
7. run `which yarn` (should fail)
8. run `corepack enable` 
9. run `which yarn` should work 🥳 

I have added an extra test case testing this scenario :)

Reply:
Ah nvm, I was thinking about the scenario where your node installation was previously cached, however we always go through the `install_node` function which does the `corepack enable` step 👍